source/common/logger/ur_sinks.h

```diff
@@ -0,0 +1,77 @@
+// Copyright (C) 2023 Intel Corporation
+// SPDX-License-Identifier: MIT
+
+#include <iostream>
+
+namespace sink {
+
+class Sink {
+  public:
+    template <typename... Args>
+    void log(const char *level, const char *fmt, Args &&...args) {
+        *ostream << "[" << level << "]:";
+        format(fmt, std::forward<Args &&>(args)...);
+        *ostream << "\n";
+        ostream->flush();
+    }
+
+  protected:
+    std::ostream *ostream;
+
+    Sink() = default;
+    virtual ~Sink() = 0;
+
+  private:
+    template <typename Arg> void format(Arg &&arg) { *ostream << arg; }
+
+    template <typename Arg, typename... Args>
+    void format(const char *fmt, Arg &&arg, Args &&...args) {
+        while (*fmt != '{' && *fmt != '\0') {
+            *ostream << *fmt++;
+        }
+
+        if (*fmt == '{') {
+            if (*(++fmt) != '}') {
+                throw std::runtime_error("Only empty braces are allowed!");
+            }
+            *ostream << arg;
+        }
+
+        format(++fmt, std::forward<Args &&>(args)...);
+    }
+};
+Sink::~Sink() {}
+
+class StdoutSink : public Sink {
+  public:
+    StdoutSink() { this->ostream = &std::cout; }
+
+    ~StdoutSink() = default;
+};
+
+class StderrSink : public Sink {
+  public:
+    StderrSink() { this->ostream = &std::cerr; }
+
+    ~StderrSink() = default;
+};
+
+class FileSink : public Sink {
+  public:
+    FileSink(const char *file_path) : file_path(file_path) {
+        ofstream = std::ofstream(file_path, std::ofstream::out);
+        if (ofstream.rdstate() != std::ofstream::goodbit) {
+            throw std::runtime_error(
+                std::string("Failure while opening log file: ") + file_path);
+        }
+        this->ostream = &ofstream;
+    }
+
+    ~FileSink() { ofstream.close(); }
+
+  private:
+    const char *file_path;
```

Comment:
If you want to store the `file_path` it should be `std::string`. Otherwise you're risking having a dangling pointer if the file path is created dynamically. But... do you really need to store this?

Reply:
Removed.